File: 2015/day4/deno/logic.ts
import { crypto } from "https://deno.land/std@0.209.0/crypto/mod.ts";
import { encodeHex } from "https://deno.land/std@0.209.0/encoding/hex.ts";

/**
 * Checks if hash starts with five zeros.
 * @param {string} hash MD5 hash string.
 */
function doesHashStartWithFiveZeros(hash: string): boolean {
    return hash.slice(0, 5) === "00000";
}

/**
 * Gets the lowest possible positive number for the secret key that when hashed has five leading zeros.
 * @param {string} secretKey Secret Key
 */
async function getLowestNumber(secretKey: string): Promise<number> {
    let lowestNumber = 0;
    const encoder = new TextEncoder();
    while (lowestNumber < Number.MAX_SAFE_INTEGER) {
        const hash  = await crypto.subtle.digest("MD5", encoder.encode(`${secretKey}${lowestNumber}`));
        const hashHex = encodeHex(hash);
        if (doesHashStartWithFiveZeros(hashHex)) {
            break;
        }
        ++lowestNumber;
    }
    return lowestNumber;
}

export { doesHashStartWithFiveZeros, getLowestNumber }

File: 2015/day4/ruby/logic.rb
require 'digest'

##
# Check if hash starts with five zeros.
def does_hash_start_with_five_zeros(hash)
  hash[0..4] == "00000"
end

##
# Gets the lowest possible positive number for the secret key that when hashed has five leading zeros.
def get_lowest_number(secret_key)
  lowest_number = 0

  # NOTE: I would use some kind of numeric constant for a maximum number
  # but Ruby doesn't seem to have one. Instead it'll turn them into Bignums
  # which are unlimited. In practice putting a hard cap on the maximum is safer.
  while true do
    hash = Digest::MD5.hexdigest("#{secret_key}#{lowest_number}")

    if does_hash_start_with_five_zeros(hash)
      break
    end

    lowest_number += 1
  end

  lowest_number
end
